### PR TITLE
MOBILEAPPS-2028: Added reachability check in each step of flow

### PR DESCRIPTION
**Pull request by ahmadhusseini**

### CODE REVIEW

File: ForeSee.Auth.Shared/LoginController.cs [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+
+        /// <summary>
+        /// Exchanges the intercepted callback code for an Access Token.
+        /// </summary>
+        /// <param name="code">The intercepted callback code.</param>
+        void BeginAccessTokenExchange(string code)
+        {
+            if (!Reachable)
+            {
+                RunOnMainThread(() => {
```
Comment (swuforesee) [inline ForeSee.Auth.Shared/LoginController.cs]: Good enforcement here to make this call on the main ui thread :thumbsup:

File: ForeSee.Auth.Shared/LoginController.cs [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Informs the LoginDelegate that the user cancelled the login flow.
+        /// </summary>
+        public void UserCancelledLogin()
+        {
+            loginDelegate.EncounteredError(ApiError.InterruptedByUser, "The user cancelled the login flow.");
```
Comment (swuforesee) [inline ForeSee.Auth.Shared/LoginController.cs]: @ahmadhusseini is there any particular reason to not explicity to notify the loginDelegate on the main ui thread?

File: ForeSee.Auth.Shared/LoginController.cs [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+        }
+
+        /// <summary>
+        /// Begins the Access Token exchange after successfully intercepting the callback and dismissing the web view.
+        /// </summary>
+        public void RetrievedAuthCode()
+        {
+            if (string.IsNullOrEmpty(authCode))
+            {
+                loginDelegate.EncounteredError(ApiError.InvalidCallbackParameters, "The Auth Code is invalid.");
```
Comment (swuforesee) [inline ForeSee.Auth.Shared/LoginController.cs]: @ahmadhusseini is there any particular reason to not explicity run this on the main ui thread?

File: ForeSee.Auth.Shared/LoginController.cs [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+        /// <summary>
+        /// Passes an ApiError to the LoginDelegate.
+        /// </summary>
+        /// <param name="error">An ApiError describing the error that was encountered.</param>
+        /// <param name="message">A message associated with the error</param>
+        /// <param name="code">The code associated with the error</param>
+
+        public void EncounteredError(ApiError error, string message, int code = 0)
+        {
+            loginDelegate.EncounteredError(error, message, code);
```
Comment (swuforesee) [inline ForeSee.Auth.Shared/LoginController.cs]: @ahmadhusseini is there any particular reason to not explicity run this on the main ui thread?

**Comment by swuforesee** [comment]
**Login library Code review notes:**
@ahmadhusseini  Changes look good. 
Noticed a couple of places where calls to the LoginDelegate are not explictly run on the main ui thread. Could you kindly take a look?

Passing to Selwn to review in the mean time.

File: ForeSee.Auth.Shared/LoginController.cs (hunk 4 above)
Comment (ahmadhusseini) [reply ForeSee.Auth.Shared/LoginController.cs]: This is a delegate method for the ExternalFlowViewController, it's called on the main thread.

File: ForeSee.Auth.Shared/LoginController.cs (hunk 2 above)
Comment (ahmadhusseini) [reply ForeSee.Auth.Shared/LoginController.cs]: This is a delegate method for the ExternalFlowViewController, it's called on the main thread.

File: ForeSee.Auth.Shared/LoginController.cs (hunk 3 above)
Comment (ahmadhusseini) [reply ForeSee.Auth.Shared/LoginController.cs]: This is a delegate method for the ExternalFlowViewController, it's called on the main thread.

**Pull request merged**